custom_components/tesla_custom/climate.py

```diff
@@ -52,125 +73,122 @@ def hvac_mode(self):
+            return self._car.min_avail_temp
+
+        return DEFAULT_MIN_TEMP
 
     @property
     def target_temperature(self):
-        """Return the temperature we try to reach."""
-        return self.tesla_device.get_goal_temp()
+        """Return target temperature."""
+        return self._car.driver_temp_setting
 
-    @TeslaDevice.Decorators.check_for_reauth
     async def async_set_temperature(self, **kwargs):
-        """Set new target temperatures."""
+        """Set new target temperature."""
         temperature = kwargs.get(ATTR_TEMPERATURE)
         if temperature:
             _LOGGER.debug("%s: Setting temperature to %s", self.name, temperature)
-            await self.tesla_device.set_temperature(temperature)
-            self.async_write_ha_state()
 
-    @TeslaDevice.Decorators.check_for_reauth
+            temp = round(temperature, 1)
+
+            await self._car.set_temperature(temp)
+            # We'll create a non-blocking update call so we don't hold up
+            # the current call.
+            await self.update_controller(
+                force=True, wake_if_asleep=True, blocking=False
+            )
+
     async def async_set_hvac_mode(self, hvac_mode):
         """Set new target hvac mode."""
         _LOGGER.debug("%s: Setting hvac mode to %s", self.name, hvac_mode)
         if hvac_mode == HVAC_MODE_OFF:
-            await self.tesla_device.set_status(False)
+            await self._car.set_hvac_mode("off")
         elif hvac_mode == HVAC_MODE_HEAT_COOL:
-            await self.tesla_device.set_status(True)
-        await self.update_climate_related_devices()
-        self.async_write_ha_state()
+            await self._car.set_hvac_mode("on")
 
-    @TeslaDevice.Decorators.check_for_reauth
-    async def async_set_preset_mode(self, preset_mode: str) -> None:
-        """Set new preset mode."""
-        _LOGGER.debug("%s: Setting preset_mode to: %s", self.name, preset_mode)
-        try:
-            await self.tesla_device.set_preset_mode(preset_mode)
-            self.async_write_ha_state()
-        except UnknownPresetMode as ex:
-            _LOGGER.error("%s", ex.message)
+        # Changing the HVAC mode can change alot of climate parms
+        # So we'll do a blocking update.
+        await self.update_controller(force=True, wake_if_asleep=True)
 
     @property
     def preset_mode(self) -> str | None:
         """Return the current preset mode, e.g., home, away, temp.
 
         Requires SUPPORT_PRESET_MODE.
         """
-        return self.tesla_device.preset_mode
+        if self._car.defrost_mode == 2:
+            return "Defrost"
+        if self._car.climate_keeper_mode == "dog":
+            return "Dog Mode"
+        if self._car.climate_keeper_mode == "camp":
+            return "Camp Mode"
+        if self._car.climate_keeper_mode == "on":
+            return "Keep On"
+
+        return "Normal"
 
     @property
     def preset_modes(self) -> list[str] | None:
```

Comment:
I don't know why but this is now throwing an error in tests

````
  File "/Users/my_user/Projects/tesla/custom_components/tesla_custom/climate.py", line 151, in TeslaCarClimate
    def preset_mode(self) -> str | None:
TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
````

It was passing before... removing the typing `list[str] | None` "fixes" it, but I'd like to figure out what's causing this to throw a type error sometimes. I'm guessing it has something to do with my virtual environment. I created a venv which is using Python 3.9 and within that venv, I did a `poetry install` and that's how I'm running these tests.